c/compress-io.c

```diff
@@ -236,334 +284,389 @@ glzFile glzdopen_input(INT fd) {
       }
+    } else {
+      /* EOF on read */
+      break;
     }
+  }
 
-    if (lz4->out_pos < lz4->out_len) {
-      UINT amt = lz4->out_len - lz4->out_pos;
-      if (amt > count) amt = count;
-      memcpy(buffer, (char *)lz4->out_buffer + lz4->out_pos, amt);
-      lz4->out_pos += amt;
-      lz4->stream_pos += amt;
-      return amt;
-    }
+  if (lz4->out_pos < lz4->out_len) {
+    UINT amt = lz4->out_len - lz4->out_pos;
+    if (amt > count) amt = count;
+    memcpy(buffer, lz4->out_buffer + lz4->out_pos, amt);
+    lz4->out_pos += amt;
+    lz4->stream_pos += amt;
+    return amt;
+  }
 
-    return 0;
+  return 0;
+}
+
+INT S_glzread(glzFile glz, void *buffer, UINT count) {
+  switch (glz->format) {
+    case COMPRESS_GZIP:
+      return gzread(glz->gz, buffer, count);
+    case COMPRESS_LZ4:
+      return glzread_lz4(glz->lz4_in, buffer, count);
+    default:
+      S_error1("S_glzread", "unexpected compress format ~s", Sinteger(glz->format));
+      return -1;
   }
 }
 
-INT glzwrite(glzFile file, void *buffer, UINT count) {
-  if (file->mode == is_gz)
-    return gzwrite(file->gz, buffer, count);
-  else {
-    lz4File_out *lz4 = (lz4File_out *)file->lz4;
-
-    if ((lz4->in_pos == USE_LZ4_BUFFER_SIZE)
-        || ((lz4->in_pos > 0) && !buffer)) {
-      size_t out_len;
-
-      out_len = LZ4F_compressFrame(lz4->out_buffer, lz4->out_buffer_size,
-                                   lz4->in_buffer, lz4->in_pos,
-                                   &lz4->preferences);
-      if (LZ4F_isError(out_len)) {
-        lz4->err = (INT)out_len;
-        return -1;
-      }
+static INT glzemit_lz4(lz4File_out *lz4, void *buffer, UINT count) {
+  ptr tc = get_thread_context();
+  sized_buffer *cached_out_buffer;
+  char *out_buffer;
+  size_t out_len, out_pos;
+  INT r = 0;
+
+  /* allocate one out_buffer (per thread) since we don't need one for each file.
+     the buffer is freed by destroy_thread. */
+  if ((cached_out_buffer = LZ4OUTBUFFER(tc)) == NULL || cached_out_buffer->size < lz4->out_buffer_size) {
+    if (cached_out_buffer != NULL) free(cached_out_buffer);
+    if ((LZ4OUTBUFFER(tc) = cached_out_buffer = malloc(sizeof(sized_buffer) + lz4->out_buffer_size)) == NULL) return -1;
+    cached_out_buffer->size = lz4->out_buffer_size;
+  }
+  out_buffer = cached_out_buffer->buffer;
 
-      lz4->in_pos = 0;
-      lz4->out_len = (INT)out_len;
-      lz4->out_pos = 0;
-    }
+  out_len = LZ4F_compressFrame(out_buffer, lz4->out_buffer_size,
+                               buffer, count,
+                               &lz4->preferences);
+  if (LZ4F_isError(out_len)) {
+    lz4->err = Z_STREAM_ERROR;
+    return -1;
+  }
 
-    while (lz4->out_pos < lz4->out_len) {
-      INT r = WIN32_IZE(write)(lz4->fd, (char*)lz4->out_buffer + lz4->out_pos, lz4->out_len - lz4->out_pos);
-      if (r >= 0)
-        lz4->out_pos += r;
+  out_pos = 0;
+  while (out_pos < out_len) {
+    r = WIN32_IZE(write)(lz4->fd, out_buffer + out_pos, out_len - out_pos);
```

Comment:
The WIN32 definition of `_write` uses `unsigned int` for the count, not `size_t`. As a result, we get a compiler warning. Here's one way to fix it:

<pre>#ifdef WIN32
    r = _write(lz4->fd, out_buffer + out_pos, (UINT)(out_len - out_pos));
#else
    r = write(lz4->fd, out_buffer + out_pos, out_len - out_pos);
#endif</pre>

Reply:
Thanks!  I ended up fixing this one by using INT for the lengths and positions.